In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_pickle('dataframes/df_v2.pkl')

In [3]:
df.head()

SUBJECT_ID  HADM_ID                                               TEXT  \
0           3   145834  sex:  m    medicine    admitted from rehabilit...   
1           4   185777  sex:  f      shortness of breath and fevers.  ...   
2           6   107064  sex:  f     admission diagnosis:  end stage re...   
3           9   150750  sex:  m    neurology    weakness, inability to...   
4          10   184167  sex:  f      baby girl  is a 1,385 gram, forme...   

                                              LABELS  
0  038.9;785.59;584.9;427.5;410.71;428.0;682.6;42...  
1  042;136.3;799.4;276.3;790.7;571.5;041.11;V09.0...  
2  403.91;444.0;997.2;276.6;276.7;285.9;275.3;V15...  
3      431;507.0;428.0;584.9;276.5;401.9;96.72;96.04  
4  V30.00;774.2;765.25;765.15;V29.0;99.83;99.15;96.6

In [4]:
import collections

def flatten(l):
    for el in l:
        if isinstance(el, collections.Iterable) and not isinstance(el, str):
            for sub in flatten(el):
                yield sub
        else:
            yield el

## Get top 50 labels

In [5]:
from collections import Counter

In [6]:
df['LABELS'] = df['LABELS'].apply(lambda x: x.split(';'))

In [7]:
df['LABELS'] = df['LABELS'].apply(lambda x: list(x))

In [8]:
labels = list(df['LABELS'])

In [9]:
len(labels)

52726

In [10]:
len(list(flatten(labels)))

848692

In [11]:
for i, row in enumerate(labels):
    for j, code in enumerate(row):
        if code == 'nan':
            del row[j]

In [12]:
len(labels)

52726

In [13]:
len(list(flatten(labels)))

848688

In [14]:
df['LABELS'] = labels

In [15]:
df['LABELS'] = df['LABELS'].apply(lambda x: set(x))
labels = list(df['LABELS'])

In [16]:
all_labels = []

for label_set in labels:
    for label in label_set:
        all_labels.append(label)
    

In [17]:
len(all_labels)

837374

In [18]:
labels_count = Counter(all_labels)

In [19]:
labels_count

Counter({'682.6': 728,
         '584.9': 8906,
         '410.71': 3001,
         '425.4': 1666,
         '785.59': 505,
         '263.9': 1240,
         '99.62': 760,
         '428.0': 12842,
         '038.9': 3580,
         '427.5': 1281,
         '89.64': 1049,
         '38.93': 12866,
         '96.6': 9102,
         '96.72': 5815,
         '96.04': 9493,
         '799.4': 245,
         '042': 522,
         '790.7': 1472,
         '136.3': 85,
         'V09.0': 655,
         'E931.7': 29,
         '276.3': 967,
         '571.5': 1250,
         '88.72': 3305,
         '041.11': 693,
         '33.23': 994,
         '39.57': 71,
         '997.2': 261,
         '275.3': 451,
         '285.9': 5295,
         '38.06': 31,
         '99.04': 7102,
         '91.': 25,
         '444.0': 54,
         '276.7': 2111,
         '276.6': 568,
         '403.91': 2566,
         'V15.82': 2741,
         '55.69': 145,
         '401.9': 20046,
         '276.5': 1294,
         '431': 1294,
         '507.0

In [20]:
# Number of unique labels

len(labels_count)

8907

In [21]:
top_50_tuples = labels_count.most_common(50)

In [22]:
top_50_tuples

[('401.9', 20046),
 ('38.93', 12866),
 ('428.0', 12842),
 ('427.31', 12589),
 ('414.01', 12178),
 ('96.04', 9493),
 ('96.6', 9102),
 ('584.9', 8906),
 ('250.00', 8783),
 ('272.4', 8503),
 ('96.71', 8426),
 ('518.81', 7249),
 ('99.04', 7102),
 ('39.61', 6781),
 ('599.0', 6442),
 ('530.81', 6154),
 ('96.72', 5815),
 ('272.0', 5766),
 ('285.9', 5295),
 ('88.56', 5045),
 ('244.9', 4785),
 ('486', 4732),
 ('285.1', 4499),
 ('38.91', 4449),
 ('36.15', 4387),
 ('276.2', 4358),
 ('496', 4296),
 ('99.15', 4162),
 ('995.92', 3792),
 ('V58.61', 3697),
 ('507.0', 3592),
 ('038.9', 3580),
 ('585.9', 3367),
 ('403.90', 3350),
 ('311', 3347),
 ('88.72', 3305),
 ('305.1', 3272),
 ('412', 3203),
 ('37.22', 3147),
 ('39.95', 3133),
 ('287.5', 3002),
 ('410.71', 3001),
 ('276.1', 2985),
 ('V45.81', 2943),
 ('424.0', 2876),
 ('V15.82', 2741),
 ('511.9', 2693),
 ('93.90', 2656),
 ('V45.82', 2651),
 ('37.23', 2619)]

In [23]:
top_50 = []

for tup in top_50_tuples:
    top_50.append(tup[0])

In [24]:
top_50 = set(top_50)

In [25]:
df_top50 = df.copy()
df_top50['LABELS'] = df_top50['LABELS'].apply(lambda x: x.intersection(top_50))

In [26]:
df_top50.head()

SUBJECT_ID  HADM_ID                                               TEXT  \
0           3   145834  sex:  m    medicine    admitted from rehabilit...   
1           4   185777  sex:  f      shortness of breath and fevers.  ...   
2           6   107064  sex:  f     admission diagnosis:  end stage re...   
3           9   150750  sex:  m    neurology    weakness, inability to...   
4          10   184167  sex:  f      baby girl  is a 1,385 gram, forme...   

                                              LABELS  
0  {584.9, 410.71, 428.0, 38.93, 038.9, 96.6, 96....  
1                                     {88.72, 38.93}  
2                             {99.04, V15.82, 285.9}  
3         {584.9, 401.9, 428.0, 96.72, 507.0, 96.04}  
4                                      {99.15, 96.6}

In [27]:
len(df_top50)

52726

In [28]:
drop_indices = (df_top50[df_top50['LABELS'].map(len) == 0]).index
df_top50 = df_top50.drop(drop_indices)

In [29]:
len(drop_indices)

3195

In [30]:
df_top50 = df_top50.reset_index(drop=True)

In [31]:
len(df_top50)

49531

In [32]:
len(df)

52726

In [33]:
df.head()

SUBJECT_ID  HADM_ID                                               TEXT  \
0           3   145834  sex:  m    medicine    admitted from rehabilit...   
1           4   185777  sex:  f      shortness of breath and fevers.  ...   
2           6   107064  sex:  f     admission diagnosis:  end stage re...   
3           9   150750  sex:  m    neurology    weakness, inability to...   
4          10   184167  sex:  f      baby girl  is a 1,385 gram, forme...   

                                              LABELS  
0  {682.6, 584.9, 410.71, 425.4, 785.59, 263.9, 9...  
1  {799.4, 042, 790.7, 136.3, V09.0, E931.7, 276....  
2  {39.57, 997.2, 275.3, 285.9, 38.06, 99.04, 91....  
3  {584.9, 401.9, 428.0, 276.5, 431, 96.72, 507.0...  
4  {774.2, 99.15, 99.83, V30.00, V29.0, 765.15, 7...

## One-hot encode

In [34]:
df_top50['LABELS'] = df_top50['LABELS'].apply(lambda x: list(x))
df['LABELS'] = df['LABELS'].apply(lambda x: list(x))

In [35]:
from sklearn.preprocessing import MultiLabelBinarizer

mlb = MultiLabelBinarizer()

df = df.join(pd.DataFrame(mlb.fit_transform(df.pop('LABELS')),
                          columns=mlb.classes_,
                          index=df.index))

df_top50 = df_top50.join(pd.DataFrame(mlb.fit_transform(df_top50.pop('LABELS')),
                                     columns=mlb.classes_,
                                     index=df_top50.index))

In [36]:
df.head()

SUBJECT_ID  HADM_ID                                               TEXT  \
0           3   145834  sex:  m    medicine    admitted from rehabilit...   
1           4   185777  sex:  f      shortness of breath and fevers.  ...   
2           6   107064  sex:  f     admission diagnosis:  end stage re...   
3           9   150750  sex:  m    neurology    weakness, inability to...   
4          10   184167  sex:  f      baby girl  is a 1,385 gram, forme...   

   003.0  003.1  003.8  003.9  004.1  004.8  004.9   ...    V87.45  V88.01  \
0      0      0      0      0      0      0      0   ...         0       0   
1      0      0      0      0      0      0      0   ...         0       0   
2      0      0      0      0      0      0      0   ...         0       0   
3      0      0      0      0      0      0      0   ...         0       0   
4      0      0      0      0      0      0      0   ...         0       0   

   V88.11  V88.12  V88.21  V90.10  V90.39  V90.81  V90.89  V91.03  
0       0       0       0       0       0       0       0       0  
1       0       0       0       0       0       0       0       0  
2       0       0       0       0       0       0       0       0  
3       0       0       0       0       0       0       0       0  
4       0       0       0       0       0       0       0       0  

[5 rows x 8910 columns]

In [37]:
df_top50.head()

SUBJECT_ID  HADM_ID                                               TEXT  \
0           3   145834  sex:  m    medicine    admitted from rehabilit...   
1           4   185777  sex:  f      shortness of breath and fevers.  ...   
2           6   107064  sex:  f     admission diagnosis:  end stage re...   
3           9   150750  sex:  m    neurology    weakness, inability to...   
4          10   184167  sex:  f      baby girl  is a 1,385 gram, forme...   

   038.9  244.9  250.00  272.0  272.4  276.1  276.2   ...    96.6  96.71  \
0      1      0       0      0      0      0      0   ...       1      0   
1      0      0       0      0      0      0      0   ...       0      0   
2      0      0       0      0      0      0      0   ...       0      0   
3      0      0       0      0      0      0      0   ...       0      0   
4      0      0       0      0      0      0      0   ...       1      0   

   96.72  99.04  99.15  995.92  V15.82  V45.81  V45.82  V58.61  
0      1      0      0       0       0       0       0       0  
1      0      0      0       0       0       0       0       0  
2      0      1      0       0       1       0       0       0  
3      1      0      0       0       0       0       0       0  
4      0      0      1       0       0       0       0       0  

[5 rows x 53 columns]

## Train/test split

In [38]:
def make_guid(df):
    guid_col = df.apply(lambda x: str(x['SUBJECT_ID']) + ',' + str(x['HADM_ID']), axis=1)
    df.insert(0, 'guid', guid_col)

In [39]:
make_guid(df)
make_guid(df_top50)

In [40]:
df.head()

guid  SUBJECT_ID  HADM_ID  \
0   3,145834           3   145834   
1   4,185777           4   185777   
2   6,107064           6   107064   
3   9,150750           9   150750   
4  10,184167          10   184167   

                                                TEXT  003.0  003.1  003.8  \
0  sex:  m    medicine    admitted from rehabilit...      0      0      0   
1  sex:  f      shortness of breath and fevers.  ...      0      0      0   
2  sex:  f     admission diagnosis:  end stage re...      0      0      0   
3  sex:  m    neurology    weakness, inability to...      0      0      0   
4  sex:  f      baby girl  is a 1,385 gram, forme...      0      0      0   

   003.9  004.1  004.8   ...    V87.45  V88.01  V88.11  V88.12  V88.21  \
0      0      0      0   ...         0       0       0       0       0   
1      0      0      0   ...         0       0       0       0       0   
2      0      0      0   ...         0       0       0       0       0   
3      0      0      0   ...         0       0       0       0       0   
4      0      0      0   ...         0       0       0       0       0   

   V90.10  V90.39  V90.81  V90.89  V91.03  
0       0       0       0       0       0  
1       0       0       0       0       0  
2       0       0       0       0       0  
3       0       0       0       0       0  
4       0       0       0       0       0  

[5 rows x 8911 columns]

In [41]:
df_top50.head()

guid  SUBJECT_ID  HADM_ID  \
0   3,145834           3   145834   
1   4,185777           4   185777   
2   6,107064           6   107064   
3   9,150750           9   150750   
4  10,184167          10   184167   

                                                TEXT  038.9  244.9  250.00  \
0  sex:  m    medicine    admitted from rehabilit...      1      0       0   
1  sex:  f      shortness of breath and fevers.  ...      0      0       0   
2  sex:  f     admission diagnosis:  end stage re...      0      0       0   
3  sex:  m    neurology    weakness, inability to...      0      0       0   
4  sex:  f      baby girl  is a 1,385 gram, forme...      0      0       0   

   272.0  272.4  276.1   ...    96.6  96.71  96.72  99.04  99.15  995.92  \
0      0      0      0   ...       1      0      1      0      0       0   
1      0      0      0   ...       0      0      0      0      0       0   
2      0      0      0   ...       0      0      0      1      0       0   
3      0      0      0   ...       0      0      1      0      0       0   
4      0      0      0   ...       1      0      0      0      1       0   

   V15.82  V45.81  V45.82  V58.61  
0       0       0       0       0  
1       0       0       0       0  
2       1       0       0       0  
3       0       0       0       0  
4       0       0       0       0  

[5 rows x 54 columns]

In [43]:
# Number of Patients in all label set.
df['SUBJECT_ID'].nunique()

41127

In [44]:
# Number of admissions in all label set.
len(df)

52726

In [45]:
# Number of patients in top 50 label set.
df_top50['SUBJECT_ID'].nunique()

38428

In [46]:
# Number of admissions in top 50 label set.
len(df_top50)

49531

In [50]:
# Number of admission in which a top 50 label was not included:
len(drop_indices)

3195

In [51]:
# Number of patients that have been dropped now after filtering to top 50:
df['SUBJECT_ID'].nunique() - df_top50['SUBJECT_ID'].nunique()

2699

So we see that 2699 patients are dropped after removing entries which do not include a top 50 label set (3195 in total). This means that repeat patients generally have codes that belong to the common set or else we wouldn't be dropping so many patients. For example, if we lost all 3195 patients, that means these were all unique patients with a single visit (since we're dropping on a row by row basis and this would mean that this patient is no longer in the dataset). 

Anyways, this is interesting to keep note of. It seems that rare codes tend to belong to patients that visit only once.

In [53]:
# Now we make train, val, test splits using predefined splits via CAML code repo.

In [60]:
train_full_hadm_ids = pd.read_csv('data_prebert/train_full_hadm_ids.csv', header=None)
dev_full_hadm_ids = pd.read_csv('data_prebert/dev_full_hadm_ids.csv', header=None)
test_full_hadm_ids = pd.read_csv('data_prebert/test_full_hadm_ids.csv', header=None)

In [61]:
train_50_hadm_ids = pd.read_csv('data_prebert/train_50_hadm_ids.csv', header=None)
dev_50_hadm_ids = pd.read_csv('data_prebert/dev_50_hadm_ids.csv', header=None)
test_50_hadm_ids = pd.read_csv('data_prebert/test_50_hadm_ids.csv', header=None)

In [92]:
frames = [train_50_hadm_ids, dev_50_hadm_ids, test_50_hadm_ids]
full_50_hadm_ids = pd.concat(frames)

In [94]:
len(full_50_hadm_ids)

11368

In [80]:
df_train_full = df[df['HADM_ID'].isin(list(flatten(train_full_hadm_ids.values)))]
df_dev_full = df[df['HADM_ID'].isin(list(flatten(dev_full_hadm_ids.values)))]
df_test_full = df[df['HADM_ID'].isin(list(flatten(test_full_hadm_ids.values)))]

In [83]:
len(df_train_full), len(df_dev_full), len(df_test_full)

(47723, 1631, 3372)

In [86]:
df_train_50 = df_top50[df_top50['HADM_ID'].isin(list(flatten(train_50_hadm_ids.values)))]
df_dev_50 = df_top50[df_top50['HADM_ID'].isin(list(flatten(dev_50_hadm_ids.values)))]
df_test_50 = df_top50[df_top50['HADM_ID'].isin(list(flatten(test_50_hadm_ids.values)))]

In [87]:
len(df_train_50), len(df_dev_50), len(df_test_50)

(8052, 1569, 1725)

In [115]:
df_train_full = df_train_full.drop(columns=['SUBJECT_ID', 'HADM_ID'])

In [117]:
df_dev_full = df_dev_full.drop(columns=['SUBJECT_ID', 'HADM_ID'])
df_test_full = df_test_full.drop(columns=['SUBJECT_ID', 'HADM_ID'])

In [118]:
df_train_50 = df_train_50.drop(columns=['SUBJECT_ID', 'HADM_ID'])
df_dev_50 = df_dev_50.drop(columns=['SUBJECT_ID', 'HADM_ID'])
df_test_50 = df_test_50.drop(columns=['SUBJECT_ID', 'HADM_ID'])

In [120]:
# Our splits are ready. Save to disk

df_train_full.to_csv('data_prebert/train_full.csv', index=False)
df_dev_full.to_csv('data_prebert/val_full.csv', index=False)
df_test_full.to_csv('data_prebert/test_full.csv', index=False)

df_train_50.to_csv('data_prebert/train_50.csv', index=False)
df_dev_50.to_csv('data_prebert/val_50.csv', index=False)
df_test_50.to_csv('data_prebert/test_50.csv', index=False)

In [125]:
', '.join((list(df_train_50.columns[2:])))

'038.9, 244.9, 250.00, 272.0, 272.4, 276.1, 276.2, 285.1, 285.9, 287.5, 305.1, 311, 36.15, 37.22, 37.23, 38.91, 38.93, 39.61, 39.95, 401.9, 403.90, 410.71, 412, 414.01, 424.0, 427.31, 428.0, 486, 496, 507.0, 511.9, 518.81, 530.81, 584.9, 585.9, 599.0, 88.56, 88.72, 93.90, 96.04, 96.6, 96.71, 96.72, 99.04, 99.15, 995.92, V15.82, V45.81, V45.82, V58.61'

In [126]:
', '.join((list(df_train_full.columns[2:])))

'003.0, 003.1, 003.8, 003.9, 004.1, 004.8, 004.9, 005.1, 005.81, 005.9, 007.1, 007.4, 008.04, 008.41, 008.43, 008.45, 008.47, 008.5, 008.61, 008.62, 008.63, 008.69, 008.8, 009.0, 009.1, 009.2, 009.3, 010.85, 011.23, 011.36, 011.64, 011.86, 011.90, 011.93, 011.94, 012.05, 012.15, 013.00, 013.04, 013.25, 013.30, 013.54, 014.02, 014.05, 014.85, 015.04, 015.05, 018.03, 018.05, 018.80, 018.90, 018.94, 018.95, 018.96, 021.8, 023.9, 027.0, 027.2, 027.9, 030.9, 031.0, 031.1, 031.2, 031.8, 031.9, 032.85, 032.89, 032.9, 033.8, 034.0, 035, 036.0, 036.2, 036.41, 038.0, 038.10, 038.11, 038.12, 038.19, 038.2, 038.3, 038.40, 038.41, 038.42, 038.43, 038.44, 038.49, 038.8, 038.9, 039.1, 039.2, 039.8, 040.0, 040.82, 040.89, 041.00, 041.01, 041.02, 041.03, 041.04, 041.05, 041.09, 041.10, 041.11, 041.12, 041.19, 041.2, 041.3, 041.4, 041.49, 041.5, 041.6, 041.7, 041.82, 041.83, 041.84, 041.85, 041.86, 041.89, 041.9, 042, 045.90, 046.3, 046.71, 047.0, 047.8, 047.9, 048, 049.0, 049.1, 049.8, 049.9, 052.0, 05